PIMS-1220, PIMS-1221, PIMS-1222

devinleighsmith opened the pull request:
Automap Evaluations
Move logic from controller to DAL

devinleighsmith commented on backend/dal/Services/Concrete/ParcelService.cs:
```diff
@@ -200,25 +201,106 @@ public Parcel Add(Parcel parcel)
-            entity.UpdatedOn = DateTime.UtcNow;
-
-            this.Context.Parcels.Update(entity);
+            //Add/Update a parcel and all child collections
```
this entire 60ish block could be replaced by:
context.Update(rootEntity);
context.SaveChanges();
which handles "Upserting" the parent and all child entities, supporting mixing and matching adding new entities and updating existing entities. The problem is that "Upserting" does not support deletes. My original thought was that I could messily handle the deletes by loading the existing parcel and then disposing so I could use Context.Update. However, the concept of detaching entities with their children is not supported as of this time. https://github.com/dotnet/efcore/issues/18406

Only solution at this time is to not handle any kinds of delete operations in this update function.

Fosol replied:
I think not supporting delete here makes the most sense presently.

devinleighsmith commented on backend/api/Controllers/ParcelController.cs:
```diff
@@ -67,6 +67,7 @@ public IActionResult GetParcels()
         /// <returns></returns>
         [HttpPost]
         [HasPermission(Permissions.PropertyView)]
+        [Route("/api/parcels/filter")]
```
Can't have two posts with the same route. @Fosol let me know if this is a breaking change.

Fosol replied:
Should be fine, these aren't used yet.

asanchezr replied:
I understand that there can be no two POST on the same route. But with this change it is now confusing to the API consumer how to filter parcels. There's now two (different) routes to do it:

- GET `/api/parcels?field1=value1&field2=value2`
- POST `/api/parcels/filter/` with a body of  `{ field1: value1, ...}`

I know the POST is there for convenience but I rather remove the [HttpPost] attribute and let it be a GET only operation. Note that I'm not saying we discard the whole `GetParcels(filter)` function, just not exposing it via HTTP

asanchezr replied:
There's a lot of debate out there on what is "right", I only ask we remain consistent, ideas?

- https://stackoverflow.com/questions/19637459/rest-api-using-post-instead-of-get
- https://stackoverflow.com/questions/46585/when-do-you-use-post-and-when-do-you-use-get
- https://www.moesif.com/blog/technical/api-design/REST-API-Design-Filtering-Sorting-and-Pagination/

devinleighsmith replied:
I think this is outside the scope of this review since I think you had the same issue with the GET/POST before I was forced to rename it. Feel free to discuss that here but I'm not planning on changing anything at this time.

Fosol commented on backend/dal/PIMSContext.cs:
```diff
@@ -103,6 +109,31 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
+            // get entries that are being Added or Updated
+            var modifiedEntries = ChangeTracker.Entries()
+                    .Where(x => (x.State == EntityState.Added || x.State == EntityState.Modified));
+            var userId = _httpContextAccessor.HttpContext.User.FindFirst(ClaimTypes.NameIdentifier).Value;
```
Do you have access to the extension method `User.GetUserId()`?

devinleighsmith replied:
oh, right. let me see.

Fosol commented on backend/dal/PIMSContext.cs:
```diff
@@ -103,6 +109,31 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
+                        entity.CreatedById = new Guid(userId);
+                        entity.CreatedOn = DateTime.UtcNow;
+                    }
+                    entity.UpdatedById = new Guid(userId);
```
Don't do this if it's a `EntityState.Delete` as it will most likely confuse EF with related properties.

devinleighsmith replied:
sure, updating.

Fosol commented on backend/dal/Services/Concrete/ParcelService.cs:
```diff
@@ -200,25 +201,106 @@ public Parcel Add(Parcel parcel)
+                .Include(p => p.Evaluations)
+                .Include(p => p.Buildings).ThenInclude(b => b.Evaluations)
+                .Include(p => p.Buildings).ThenInclude(b => b.Address)
+                .SingleOrDefault(u => u.Id == parcel.Id) ?? throw new KeyNotFoundException();
```
Probably should change to `FirstOrDefault(...)`

devinleighsmith replied:
Hmm, I don't know, I kind of like the documented behaviour of this when we are searching for a unique key:

Returns the only element of a sequence, or a default value if the sequence is empty; this method throws an exception if there is more than one element in the sequence.

seems appropriate since there should never be more than one match.

Fosol replied:
At the end of the day it doesn't change much here.  As the LINQ statement becomes an SQL statement and if it returns more than one it would throw an error.  However since you're making a request for the primary key there never will be more than one or none returned.  So using `Single` technically is doing more work than is required.

asanchezr commented:
Couple of comments re: process. 

- The title of this PR is PIMS-976 but I can't seem to find that issue in JIRA. A more descriptive title would help me understand what I should be reviewing
- The PR "source" branch is named PIMS-1220 which doesn't match the issue number of the PR title so I'm a bit lost what this PR is about. PIMS-1220 is about single parcel GET and I see a lot of changes related to creating, updating and deleting parcels...

devinleighsmith commented:
@asanchezr Agreed. I messed up the commit message on this one, I'm not in any rush so I'll correct it.

Also I ended up doing all three API stories at the same time. Theoretically I'd rather have submitted a PR for each one separately. However, realistically I updated all three at the same time as
1) I was using the API outputs to test the inputs, so as I updated one API I often found/resolved issues in the others.
2) There were some foundational changes required by all three APIs.

I'm also really annoyed that Github doesn't allow me to respond directly to your comment above. Why Github, Why?
